### PR TITLE
Supporting OSX

### PR DESCRIPTION
**Pull request by AngLi-Leon**
Todo/note:

1. ~~I temporally disabled `Werror` flag for clang, we may want to turn it on later.~~ Warnings that I have to temporally disable ([Full log](https://drive.google.com/open?id=0B9XVLkTHhsKpdHluNnl6eXpLUGs)):  
- `-Wtautological-undefined-compare` (happens once, removing nullptr check will result in failed test: "ERROR: AddressSanitizer: SEGV on unknown address" in Linux)
/Users/angli/Documents/peloton_mac/peloton/src/include/optimizer/property.h:52:9: error: 
      'this' pointer cannot be null in well-defined C++ code; comparison may be
      assumed to always evaluate to true
      [-Werror,-Wtautological-undefined-compare]
    if (this != nullptr && typeid(*this) == typeid(T)) {
        ^~~~    ~~~~~~~

- `-Wbraced-scalar-init` (Xcode library not compatible)
/Users/angli/Documents/peloton_mac/peloton/src/include/common/thread_pool.h:37:29: error: 
      braces around scalar initializer [-Werror,-Wbraced-scalar-init]
    current_thread_count_ = ATOMIC_VAR_INIT(0);
                            ^~~~~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/atomic:1780:30: note: 
      expanded from macro 'ATOMIC_VAR_INIT'
#define ATOMIC_VAR_INIT(__v) {__v}
                             ^~~~~

- ~~`-Wdeprecated-register` (deprecated keyword)
/Users/angli/Documents/peloton_mac/peloton/src/include/index/bloom_filter.h:141:5: error: 
      'register' storage class specifier is deprecated and incompatible with
      C++1z [-Werror,-Wdeprecated-register]
    register size_t hash_value = value_hash_obj(value);
    ^~~~~~~~~~~

- ~~`-Wunused-const-variable` (third party library...)
/Users/angli/Documents/peloton_mac/peloton/third_party/gmock/gmock-gtest-all.cc:7921:12: error: 
      unused variable 'kPathSeparatorString' [-Werror,-Wunused-const-variable]
const char kPathSeparatorString[] = "/";~~
           ^

- `-Wunused-private-field` (I have tried to use UNUSED_ATTRIBUTE macro, it works for OSX but failed for linux build)
/Users/angli/Documents/peloton_mac/peloton/src/include/common/exception.h:200:17: error: 
      private field 'type' is not used [-Werror,-Wunused-private-field]
  ExceptionType type;
                ^

- `-Wconstant-conversion` (a few places, type sizes are different in clang)
/Users/angli/Documents/peloton_mac/peloton/src/include/index/compact_ints_key.h:199:46: error: 
      implicit conversion from 'int' to 'short' changes value from 32768 to
      -32768 [-Werror,-Wconstant-conversion]
    IntType mask = static_cast<IntType>(0x1) << (sizeof(IntType) * 8UL - 1);
            ~~~~   ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- `-Wpotentially-evaluated-expression` (not sure, I can take a look at it later)
/Users/angli/Documents/peloton_mac/peloton/src/include/optimizer/operator_node.h:150:24: error: 
      expression with side effects will be evaluated despite being used as an
      operand to 'typeid' [-Werror,-Wpotentially-evaluated-expression]
    if (node && typeid(*node) == typeid(T)) {
                       ^

- `-Winfinite-recursion` (We call GetInfo from itself but there is a for-statement outside this calling place)
/Users/angli/Documents/peloton_mac/peloton/src/parser/abstract_parse.cpp:47:50: error: 
      all paths through this function will call itself
      [-Werror,-Winfinite-recursion]
const std::string AbstractParse::GetInfo() const {
                                                 ^
2. ~~Package installation script~~
3. ~~`RUSAGE_THREAD` is Linux-specific, we may want to find its equivalent.~~ For now I use `thread_info` from `thread_act.h` to get cpu usage per thread.
4. ~~Codegen support (fixing test).~~ For now use `ASAN_OPTIONS=detect_container_overflow=0` under osx.
There are two places causing the container overflow, one is related to codegen, the other is related to the libevent library.
5. ~~There is a pinned thread method in our benchmark. AFAIK, MacOS doesn't implement processor affinity for processes nor threads (https://superuser.com/questions/149312/how-to-set-processor-affinity-on-os-x).~~ For now disable pinned thread for benchmark under osx.
#504

### CODE REVIEW

**Comment by AngLi-Leon** [comment]
@apavlo @pmenon I have updated the example warning messages in above comment. Please take a look so that we can discuss our plan later, thanks!

**Comment by apavlo** [comment]
@AngLi-Leon Try adding OSX to travis.yml to see how this builds with clang.

File: src/include/index/bwtree.h [hunk 1]
```diff
@@ -1885,18 +1887,19 @@ class BwTree : public BwTreeBase {
...
    */
   class AllocationMeta {
    public:
     // One reasonable amount of memory for each chunk is 
     // delta chain len * struct len + sizeof this struct
-    static constexpr size_t CHUNK_SIZE = \
-      sizeof(DeltaNodeUnion) * 8 + sizeof(AllocationMeta);
-    
-   private: 
+    static constexpr size_t CHUNK_SIZE() {  // Fix for Mac incomplete type error
```
Comment (pmenon) [inline src/include/index/bwtree.h]: Remove comment.

File: src/optimizer/memo.cpp [hunk 2]
```diff
@@ -37,7 +35,7 @@ std::shared_ptr<GroupExpression> Memo::InsertExpression(
     assert(target_group == UNDEFINED_GROUP ||
            target_group == leaf->origin_group);
     gexpr->SetGroupID(leaf->origin_group);
-    return false;
+    return nullptr; // Fix for Mac support
```
Comment (pmenon) [inline src/optimizer/memo.cpp]: Remove comment.

File: src/parser/parser_utils.cpp [hunk 3]
```diff
@@ -261,9 +261,10 @@ void GetInsertStatementInfo(InsertStatement* stmt, uint num_indent) {
   }
 }
 
-void GetDeleteStatementInfo(DeleteStatement* stmt, uint num_indent) {
-  stmt = stmt;
-  num_indent = num_indent;
+void GetDeleteStatementInfo(DeleteStatement* stmt UNUSED_ATTRIBUTE, uint num_indent UNUSED_ATTRIBUTE) {
+//  stmt = stmt;
+//  num_indent = num_indent;
+  return;
```
Comment (pmenon) [inline src/parser/parser_utils.cpp]: Since this function is doing nothing, can it be removed?

File: script/installation/packages.sh [hunk 4]
```diff
@@ -156,7 +156,32 @@ elif [[ "$DISTRO" == *"REDHAT"* ]] && [[ "${DISTRO_VER%.*}" == "7" ]]; then
...
+    brew install git
+    brew install cmake
+    brew install gflags
+    brew install protobuf
+    brew install bison
+    brew install flex
+    brew install libevent
+    brew install boost
+    brew install jemalloc
+    brew install valgrind
```
Comment (seojungmin) [inline script/installation/packages.sh]: Do we still use valgrind? I wonder whether this is still required.

Comment (apavlo) [inline script/installation/packages.sh]: You're right. We should drop it.

File: script/testing/psql/psql_test.sh [hunk 5]
```diff
@@ -7,6 +7,7 @@
 # NOTE: absolute path to peloton directory is calculated from current directory
 # directory structure: peloton/scripts/testing/psql/<this_file>
 CODE_SOURCE_DIR=`dirname $0`
+OS_NAME=$1
```
Comment (apavlo) [inline script/testing/psql/psql_test.sh]: Why not get this information the same way that we get it in packages.sh?

File: src/codegen/oa_hash_table.cpp [hunk 6]
```diff
@@ -201,7 +201,7 @@ std::pair<llvm::Value *, llvm::Value *> OAHashTable::GetDataCountAndPointer(
                                  "singleValue"};
   {
     data_count_inline = codegen.Const64(1);
-    data_ptr_inline = AdvancePointer(codegen, after_key_p, 0UL);
+    data_ptr_inline = AdvancePointer(codegen, after_key_p, (uint64_t) 0UL);
```
Comment (pmenon) [inline src/codegen/oa_hash_table.cpp]: Change to `0ULL`

File: src/codegen/util/cc_hash_table.cpp [hunk 7]
```diff
@@ -79,7 +79,7 @@ char* CCHashTable::StoreTuple(uint64_t hash, uint32_t size) {
 // Clean up any resources this hash table has
 //===----------------------------------------------------------------------===//
 void CCHashTable::Destroy() {
-  LOG_DEBUG("Cleaning up hash table with %ld entries ...", num_elements_);
+  LOG_DEBUG("Cleaning up hash table with %llu entries ...", (unsigned long long) num_elements_);
```
Comment (pmenon) [inline src/codegen/util/cc_hash_table.cpp]: I think you can get rid of the cast if you're using the `%llu` format specifier.

File: src/codegen/util/cc_hash_table.cpp [hunk 8]
```diff
@@ -90,7 +90,7 @@ void CCHashTable::Destroy() {
       chain_length++;
     }
     if (chain_length > kMaxHashChainSize) {
-      LOG_DEBUG("Bucket %ld chain length = %u ...", i, chain_length);
+      LOG_DEBUG("Bucket %llu chain length = %u ...", (unsigned long long) i, chain_length);
```
Comment (pmenon) [inline src/codegen/util/cc_hash_table.cpp]: Ditto.

File: src/include/benchmark/ycsb/ycsb_workload.h [hunk 9]
```diff
@@ -41,7 +41,9 @@ std::vector<std::vector<type::Value>> ExecuteRead(executor::AbstractExecutor* ex
 
 void ExecuteUpdate(executor::AbstractExecutor* executor);
 
+#ifndef __APPLE__ // https://superuser.com/questions/149312/how-to-set-processor-affinity-on-os-x
```
Comment (pmenon) [inline src/include/benchmark/ycsb/ycsb_workload.h]: Can you add a brief instructive comment inside the if-def about why this was done? Include the link for those that want to know more.

File: src/include/type/type.h [hunk 10]
```diff
@@ -14,6 +14,7 @@
 
 #include <cstdint>
 #include <string>
+#include <functional>
```
Comment (pmenon) [inline src/include/type/type.h]: Why is this needed?

File: src/include/util/portable_endian.h [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+#define __LITTLE_ENDIAN LITTLE_ENDIAN
+#define __PDP_ENDIAN PDP_ENDIAN
+
+#else
+
+#error platform not supported
+
+#endif
+
+#endif
```
Comment (pmenon) [inline src/include/util/portable_endian.h]: @apavlo I think we're good w.r.t. licensing by including this (https://www.apache.org/legal/resolved.html#can-works-placed-in-the-public-domain-be-included-in-apache-products)

File: src/logging/log_buffer_pool.cpp [hunk 12]
```diff
@@ -24,7 +24,7 @@ namespace logging {
     size_t head_idx = head_ % buffer_queue_size_;
     while (true) {
       if (head_.load() < tail_.load() - 1) {
-        if (local_buffer_queue_[head_idx] == false) {
+        if (local_buffer_queue_[head_idx].get() == nullptr) { // Fix for Mac support
```
Comment (pmenon) [inline src/logging/log_buffer_pool.cpp]: Removing comment.

File: src/main/sdbench/sdbench_workload.cpp [hunk 13]
```diff
@@ -876,7 +876,7 @@ static void AggregateQueryHelper(const std::vector<oid_t> &tuple_key_attrs,
   oid_t tuple_idx = 1;  // tuple2
   for (col_itr = 0; col_itr < column_count; col_itr++) {
     direct_map_list.push_back({col_itr, {tuple_idx, col_itr}});
-    col_itr++;
+//    col_itr++;
```
Comment (pmenon) [inline src/main/sdbench/sdbench_workload.cpp]: Why is this commented out?

File: src/include/benchmark/tpcc/tpcc_workload.h [hunk 14]
```diff
@@ -51,7 +51,9 @@ void ExecuteUpdate(executor::AbstractExecutor* executor);
 
 void ExecuteDelete(executor::AbstractExecutor* executor);
 
-void PinToCore(size_t core);
+#ifndef __APPLE__ // https://superuser.com/questions/149312/how-to-set-processor-affinity-on-os-x
+    void PinToCore(size_t core);
+#endif
```
Comment (pmenon) [inline src/include/benchmark/tpcc/tpcc_workload.h]: Instead of commenting it out in the header, can you leave it in and pound-define it out the CPP implementation? Less noise in the code that way imho.

File: src/main/ycsb/ycsb_workload.cpp [hunk 15]
```diff
@@ -84,16 +84,20 @@ volatile bool is_running = true;
...
   pthread_setaffinity_np(pthread_self(), sizeof(cpu_set_t), &cpuset);
 }
+#endif
 
 void RunBackend(const size_t thread_id) {
-  
+
+#ifndef __APPLE__
   PinToCore(thread_id);
+#endif
```
Comment (pmenon) [inline src/main/ycsb/ycsb_workload.cpp]: Refer to previous comment. Leave `PinToCore` in, but pound-define out implementation. It'll be compiled away on OSX.

File: src/include/util/portable_endian.h (hunk 11 above)
Comment (apavlo) [inline src/include/util/portable_endian.h]: Agreed.

File: src/codegen/oa_hash_table.cpp (hunk 6 above)
Comment (AngLi-Leon) [inline src/codegen/oa_hash_table.cpp]: I have tried this, this fails in Linux (type sizes are different for two compilers):
```
/data/peloton/src/codegen/oa_hash_table.cpp: In member function ‘std::pair<llvm::Value*, llvm::Value*> peloton::codegen::OAHashTable::GetDataCountAndPointer(peloton::codegen::CodeGen&, llvm::Value*, llvm::Value*) const’:
/data/peloton/src/codegen/oa_hash_table.cpp:205:63: error: call of overloaded ‘AdvancePointer(peloton::codegen::CodeGen&, llvm::Value*&, long long unsigned int)’ is ambiguous
    data_ptr_inline = AdvancePointer(codegen, after_key_p, 0ULL);
                                                               ^
/data/peloton/src/codegen/oa_hash_table.cpp:205:63: note: candidates are:
/data/peloton/src/codegen/oa_hash_table.cpp:100:14: note: llvm::Value* peloton::codegen::OAHashTable::AdvancePointer(peloton::codegen::CodeGen&, llvm::Value*, uint64_t) const
 llvm::Value *OAHashTable::AdvancePointer(CodeGen &codegen, llvm::Value *ptr,
              ^
/data/peloton/src/codegen/oa_hash_table.cpp:89:14: note: llvm::Value* peloton::codegen::OAHashTable::AdvancePointer(peloton::codegen::CodeGen&, llvm::Value*, llvm::Value*) const
 llvm::Value *OAHashTable::AdvancePointer(CodeGen &codegen, llvm::Value *ptr,
              ^

```

File: src/codegen/util/cc_hash_table.cpp (hunk 7 above)
Comment (AngLi-Leon) [inline src/codegen/util/cc_hash_table.cpp]: Same here (linux)
```
In file included from /data/peloton/src/codegen/util/cc_hash_table.cpp:15:0:
/data/peloton/src/codegen/util/cc_hash_table.cpp: In member function ‘void peloton::codegen::util::CCHashTable::Destroy()’:
/data/peloton/src/include/common/logger.h:124:43: error: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 3 has type ‘uint64_t {aka long unsigned int}’ [-Werror=format=]
   ::fprintf(LOG_OUTPUT_STREAM, __VA_ARGS__);                                  \
                                           ^
/data/peloton/src/codegen/util/cc_hash_table.cpp:82:3: note: in expansion of macro ‘LOG_DEBUG’
   LOG_DEBUG("Cleaning up hash table with %llu entries ...", num_elements_);
   ^

```

File: src/include/benchmark/ycsb/ycsb_workload.h (hunk 9 above)
Comment (AngLi-Leon) [inline src/include/benchmark/ycsb/ycsb_workload.h]: Sure!

File: src/include/type/type.h (hunk 10 above)
Comment (AngLi-Leon) [inline src/include/type/type.h]: The std::hash (http://en.cppreference.com/w/cpp/utility/hash) in line [35](https://github.com/AngLi-Leon/peloton/blob/13fa99bf672990ce891dde92a135e31d7dfd3e61/src/include/type/type.h#L35) need this.

If I don't explicitly include it here (under Mac), it will link to another lib and get error
```
/Users/angli/Documents/peloton_mac/peloton/src/include/type/type.h:35:14: error: implicit instantiation of undefined template 'std::__1::hash<unsigned int>'
      return std::hash<uint32_t>{}(static_cast<uint32_t>(type_id));
             ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3171:29: note: template is declared here
template <class _Tp> struct hash;
                            ^
 
```

File: src/main/sdbench/sdbench_workload.cpp (hunk 13 above)
Comment (AngLi-Leon) [inline src/main/sdbench/sdbench_workload.cpp]: This will trigger a warning in clang. Actually, here col_itr is added twice per each iteration, I am wondering if it's a typo?

Comment (apavlo) [inline src/main/sdbench/sdbench_workload.cpp]: This looks like a typo. Check with @jarulraj

File: src/codegen/oa_hash_table.cpp (hunk 6 above)
Comment (pmenon) [inline src/codegen/oa_hash_table.cpp]: Fair enough.

File: src/include/type/type.h (hunk 10 above)
Comment (pmenon) [inline src/include/type/type.h]: Makes sense.

File: src/codegen/util/cc_hash_table.cpp (hunk 7 above)
Comment (AngLi-Leon) [inline src/codegen/util/cc_hash_table.cpp]: Failed in Mac 
```
/Users/angli/Documents/peloton_mac/peloton/src/codegen/util/cc_hash_table.cpp:82:60: error: 
      format specifies type 'unsigned long' but the argument has type 'uint64_t'
      (aka 'unsigned long long') [-Werror,-Wformat]
  LOG_DEBUG("Cleaning up hash table with %lu entries ...", num_elements_);
                                         ~~~               ^~~~~~~~~~~~~
                                         %llu
/Users/angli/Documents/peloton_mac/peloton/src/include/common/logger.h:124:32: note: 
      expanded from macro 'LOG_DEBUG'
  ::fprintf(LOG_OUTPUT_STREAM, __VA_ARGS__);                                  \
                               ^~~~~~~~~~~

```
The reason is that uint64_t is unsigned long long in Mac but unsigned long in Linux

**Comment by AngLi-Leon** [comment]
@apavlo Can be merged.

**Comment by rohit-cmu** [comment]
I saw this PR is merged and thus tried to build Peloton on macOS. However after running the installation script(Succesful) and cmake (Successful) , when I ran make -j4, I got the following error:-

```
error: unknown warning option '-Wno-tautological-undefined-compare'; did you mean '-Wno-tautological-compare'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-braced-scalar-init'; did you mean '-Wno-static-self-init'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-potentially-evaluated-expression' [-Werror,-Wunknown-warning-option]
```
I could see the same warning options mentioned in the TODO as well. So is there a way around this to build Peloton on macOS or for now we can't build it on macOS ?

**Comment by apavlo** [comment]
Submit a separate issue. AFAIK OSX should be working. We build it on
Travis.
